python/ql/src/experimental/semmle/python/frameworks/Werkzeug.qll

```diff
@@ -6,70 +6,120 @@ private import python
+         *
+         * See https://werkzeug.palletsprojects.com/en/1.0.x/datastructures/#werkzeug.datastructures.Headers.getlist
+         */
+        DataFlow::Node getlist() { result = getlist(DataFlow::TypeTracker::end()) }
       }
 
-      DataFlow::Node getlist() { result = getlist(DataFlow::TypeTracker::end()) }
-    }
+      /**
+       * Provides models for the `werkzeug.datastrctures.FileStorage` class
+       *
+       * See https://werkzeug.palletsprojects.com/en/1.0.x/datastructures/#werkzeug.datastructures.FileStorage.
+       */
+      module FileStorage {
+        /**
+         * A source of an instance of `werkzeug.datastrctures.FileStorage`.
+         *
+         * This can include instantiation of the class, return value from function
+         * calls, or a special parameter that will be set when functions are call by external
+         * library.
+         *
+         * Use `FileStorage::instance()` predicate to get references to instances of `werkzeug.datastrctures.FileStorage`.
+         */
+        abstract class InstanceSource extends DataFlow::Node { }
+
+        /** Gets a reference to an instance of `werkzeug.datastrctures.FileStorage`. */
+        private DataFlow::Node instance(DataFlow::TypeTracker t) {
+          t.start() and
+          result instanceof InstanceSource
+          or
+          exists(DataFlow::TypeTracker t2 | result = instance(t2).track(t2, t))
+        }
 
-    private class MultiDictAdditionalTaintStep extends TaintTracking::AdditionalTaintStep {
-      override predicate step(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
-        // obj -> obj.getlist
-        nodeTo.asCfgNode().(AttrNode).getObject("getlist") = nodeFrom.asCfgNode() and
-        nodeTo = MultiDictTracking::getlist()
-        or
-        // getlist -> getlist()
-        nodeFrom = MultiDictTracking::getlist() and
-        nodeTo.asCfgNode().(CallNode).getFunction() = nodeFrom.asCfgNode()
+        /** Gets a reference to an instance of `werkzeug.datastrctures.FileStorage`. */
+        DataFlow::Node instance() { result = instance(DataFlow::TypeTracker::end()) }
       }
     }
+  }
 
-    // ---------------------------------------------------------------------- //
-    // FileStorage                                                              //
-    // ---------------------------------------------------------------------- //
-    /**
-     * A Node representing an instance of a werkzeug.datastructures.FileStorage
-     *
-     * See https://werkzeug.palletsprojects.com/en/1.0.x/datastructures/#werkzeug.datastructures.FileStorage
-     */
-    abstract class FileStorage extends DataFlow::Node { }
+  private class MultiDictAdditionalTaintStep extends TaintTracking::AdditionalTaintStep {
+    override predicate step(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
+      // obj -> obj.getlist
+      exists(DataFlow::AttrRead read |
+        read.getObject() = nodeFrom and
+        nodeTo = read and
```

Comment:
Note to self: I think we should add an `AttrRead::accesses` method that encapsulates this idiom. "Bare" equalities like these look weird to me.